metagraph/core/dask/resolver.py

```diff
@@ -0,0 +1,185 @@
+import types
+from dask.base import tokenize
+from dask import delayed, is_dask_collection
+from ..resolver import Resolver, Namespace, PlanNamespace, Dispatcher
+from .placeholder import Placeholder, DelayedWrapper
+from ..plugin import ConcreteType, MetaWrapper
+from typing import Optional
+
+
+class DaskResolver:
+    _placeholders = {}
+
+    def __init__(self, resolver: Resolver):
+        self._resolver = resolver
+
+        # Patch plan namespace
+        self.plan = PlanNamespace(self)
+        self.plan.algos = self._resolver.plan.algos
+
+        # Patch algorithms
+        def build_algos(namespace):
+            for name in dir(namespace):
+                obj = getattr(namespace, name)
+                if isinstance(obj, Dispatcher):
+                    self.algos._register(
+                        obj._algo_name, Dispatcher(self, obj._algo_name)
+                    )
+                elif isinstance(obj, Namespace):
+                    build_algos(obj)
+
+        self.algos = Namespace()
+        build_algos(self._resolver.algos)
+
+        # Patch wrappers
+        def build_wrappers(namespace):
+            for name in dir(namespace):
+                obj = getattr(namespace, name)
+                if isinstance(obj, MetaWrapper):
+                    self.wrappers._register(
+                        f"{obj.Type.abstract.__name__}.{obj.__name__}",
+                        self.delayed_wrapper(obj, obj.Type),
+                    )
+                elif isinstance(obj, Namespace):
+                    build_wrappers(obj)
+
+        self.wrappers = Namespace()
+        build_wrappers(self._resolver.wrappers)
+
+        # Add placeholder types to `class_to_concrete`
+        self.class_to_concrete = self._resolver.class_to_concrete.copy()
+        for ct in self._resolver.concrete_types:
+            ph = self._get_placeholder(ct)
+            self.class_to_concrete[ph] = ct
+
+    # Default behavior (unless overridden) is to delegate to the original resolver
+    def __getattr__(self, item):
+        obj = getattr(self._resolver, item)
+        if type(obj) is types.MethodType and type(obj.__self__) is not type:
+            # Replace original resolver with the dask resolver for instance methods
+            return obj.__func__.__get__(self)
+        return obj
+
+    def __dir__(self):
+        return dir(self._resolver)
+
+    def delayed_wrapper(self, klass, concrete_type: Optional[ConcreteType] = None):
+        ct = concrete_type
+        if ct is None:
+            ct = self._resolver.class_to_concrete.get(klass)
+            if ct is None:
+                raise TypeError(
+                    f"{klass.__name__} is not a defined `value_type`. Must provide `concrete_type`."
+                )
+        ph = self._get_placeholder(ct)
+        return DelayedWrapper(klass, ph)
+
+    def _get_placeholder(self, concrete_type):
```

Comment:
Can you put some doc strings explaining what the internal methods on this class are doing?  I think I get it from reading the body, but this code is a little hard to follow at first.

Reply:
Done.